crates/rattler/src/validation.rs

```diff
@@ -0,0 +1,298 @@
+//! Functionality to validate the contents of a Conda package archive.
+//!
+//! All Conda packages contain a file `info/paths.json` that describes all the files the package contains.
+//! The [`validate_package_files`] function validates that a directory containing an extracted Conda
+//! package archive actually contains the files as described by the `paths.json` file.
+
+use rattler_conda_types::package::{PathType, PathsEntry, PathsJson};
+use sha2::{Digest, Sha256};
+use std::fs::{File, Metadata};
+use std::path::{Path, PathBuf};
+
+/// An error that is returned by [`validate_package_files`] if the contents of the directory seems to be
+/// corrupted.
+#[derive(Debug, thiserror::Error)]
+pub enum PackageValidationError {
+    #[error("failed to read 'paths.json' file")]
+    ReadPathsJsonError(#[source] std::io::Error),
+
+    #[error("the path '{0}' seems to be corrupted")]
+    CorruptedEntry(PathBuf, #[source] PackageEntryValidationError),
+}
+
+/// An error that indicates that a specific file in a package archive directory seems to be corrupted.
+#[derive(Debug, thiserror::Error)]
+pub enum PackageEntryValidationError {
+    #[error("failed to retrieve file metadata'")]
+    GetMetadataFailed(#[source] std::io::Error),
+
+    #[error("the file does not exist")]
+    NotFound,
+
+    #[error("expected a symbolic link")]
+    ExpectedSymlink,
+
+    #[error("expected a directory")]
+    ExpectedDirectory,
+
+    #[error("incorrect size, expected {0} but file on disk is {1}")]
+    IncorrectSize(u64, u64),
+
+    #[error("an io error occurred")]
+    IoError(#[from] std::io::Error),
+
+    #[error("sha256 hash mismatch, expected '{0}' but file on disk is '{1}'")]
+    HashMismatch(String, String),
+}
+
+/// Determine whether the files in the specified directory match what is expected according to the
+/// `info/paths.json` file in the same directory.
+pub fn validate_package_files(package_dir: &Path) -> Result<(), PackageValidationError> {
+    // Read the 'paths.json' file which describes all files that should be present
+    let paths = PathsJson::from_path(&package_dir.join("info/paths.json"))
```

Comment:
Some older packages do not have the `paths.json` entry, and only have a flat file called `files` that contains a list of all paths that are supposedly there. I think for maximum compatibility we should also implement that as a fallback if `paths.json` isn't availbe. The `files` file also doesn't contain any hashes, so any validation beyond the existence check of the file is hard :)

Reply:
I just pushed code to also check the deprecated `files` format by converting the `files` and friends to a `paths.json` file. I also added a few tests for this.